WordPress/WordPressTest/ContextManagerMock+AutomaticTeardown.swift

```diff
@@ -0,0 +1,72 @@
+import XCTest
+
+@testable import WordPress
+
+extension ContextManagerMock {
+    /// Override `ContextManager.shared` with the mock instance, and un-override it when
+    /// given `testCase` finishes.
+    ///
+    /// This method should be called during test set up, typically from `XCTestCase` subclass's
+    /// `setUp` method.
+    ///
+    /// - Parameter testCase: The test case to wait for.
+    @objc func useAsSharedInstance(untilTestFinished testCase: XCTestCase) {
+        // Create the test observer singleton to add it to `XCTestObservationCenter`.
+        _ = AutomaticTeardownTestObserver.instance
+
+        let original = ContextManager.overrideInstance
+        ContextManager.internalSharedInstance()
+        ContextManager.overrideSharedInstance(self)
+
+        // This closure is going to be called by the test observer below when
+        // the test case finishes. The reason a combination of storing a closure in
+        // `XCTestCase` instance and a dedicated `XCTestObservation` implementation
+        // is used here is, we need to make sure tearing down the `ContextManager`
+        // mock instance happens _after_ test finishes execution.
+        //
+        // XCTest does have an API allowing us to add our own teardown block:
+        // `XCTestCase.addTeardownBlock`, but the added block is called _before_
+        // the test case's `tearDown` override method. Which means if this official
+        // API is used here instead, `ContextManager.shared` references two different
+        // objects during test execution: the mock instance before `tearDown`, or the
+        // real singleton during `tearDown`, which isn't ideal.
+        testCase.additionalTeardown = { [weak self] in
+            guard let self = self else {
+                return
+            }
+            self.mainContext.reset()
+            if ContextManager.overrideInstance === self {
+                ContextManager.overrideSharedInstance(original)
+            }
+        }
+    }
+}
+
+private class AutomaticTeardownTestObserver: NSObject, XCTestObservation {
+
+    static let instance = AutomaticTeardownTestObserver()
+
+    private override init() {
+        super.init()
+        XCTestObservationCenter.shared.addTestObserver(self)
+    }
+
+    func testCaseDidFinish(_ testCase: XCTestCase) {
+        testCase.additionalTeardown?()
```

Comment:
I think you need to reset the `additionalTeardown` block once you've called it, otherwise the next test case might end up executing it, even if that next test case did not call `useAsSharedInstance(…)`.

This is because, if I'm not mistaken, the XCTest harness will (1) allocate an instance of the `XCTestCase` class (2) run `setUp`, then test 1, then `tearDown` (and `testCaseDidFinish`), (3) run `setUp`, then test 2, then `tearDown` (and `testCaseDidFinish`), (4) etc for each test (5) finally release the `XCTestCase` instance. As opposed to deallocating the `XCTestCase` instance between each test case and `test…` method invocation. At least I think that's how the cycle goes 😅 

```suggestion
        testCase.additionalTeardown?()
        testCase.additionalTeardown = nil
```

Reply:
Good catch! It's always good to make sure per-test environment is properly cleaned up, regardless how the test is run.

The steps you described were my expectation too, but I think that might not be the case. @mokagio and I happened to discuss this in another PR: https://github.com/wordpress-mobile/WordPress-iOS/pull/18527#discussion_r867342970. The result I observed was, a new XCTestCase instance is created for each test run.

Reply:
Wow, TIL! 👍  Still doesn't hurt to set it back to `nil`, but now I know my expectations were wrong about how XCTestCase's lifecycle works; seems I was not the only one: see [this article about that exact topic](https://qualitycoding.org/xctestcase-teardown/)